packages/pds/src/services/repo/blobs.ts

```diff
@@ -70,6 +80,74 @@ export class RepoBlobs {
     await Promise.all(blobPromises)
   }
 
+  async deleteDereferencedBlobs(did: string, writes: PreparedWrite[]) {
+    const deletes = writes.filter(
+      (w) => w.action === WriteOpAction.Delete,
+    ) as PreparedDelete[]
+    const updates = writes.filter(
+      (w) => w.action === WriteOpAction.Update,
+    ) as PreparedUpdate[]
+    const uris = [...deletes, ...updates].map((w) => w.uri.toString())
+    if (uris.length === 0) return
+
+    const deletedRepoBlobs = await this.db.db
+      .deleteFrom('repo_blob')
+      .where('did', '=', did)
+      .where('recordUri', 'in', uris)
+      .returningAll()
+      .execute()
+    if (deletedRepoBlobs.length < 1) return
+
+    const deletedRepoBlobCids = deletedRepoBlobs.map((row) => row.cid)
+    const duplicateCids = await this.db.db
+      .selectFrom('repo_blob')
+      .where('did', '=', did)
+      .where('cid', 'in', deletedRepoBlobCids)
+      .select('cid')
+      .execute()
+
+    const newBlobCids = writes
+      .map((w) =>
+        w.action === WriteOpAction.Create || w.action === WriteOpAction.Update
+          ? w.blobs
+          : [],
+      )
+      .flat()
+      .map((b) => b.cid.toString())
+    const cidsToKeep = [...newBlobCids, ...duplicateCids.map((row) => row.cid)]
+    const cidsToDelete = deletedRepoBlobCids.filter(
+      (cid) => !cidsToKeep.includes(cid),
+    )
+    if (cidsToDelete.length < 1) return
+
+    await this.db.db
+      .deleteFrom('blob')
+      .where('creator', '=', did)
+      .where('cid', 'in', cidsToDelete)
+      .execute()
+
+    // check if these blobs are used by other users before deleting from blobstore
+    const stillUsedRes = await this.db.db
+      .selectFrom('blob')
+      .where('cid', 'in', cidsToDelete)
+      .select('cid')
+      .distinct()
+      .execute()
+    const stillUsed = stillUsedRes.map((row) => row.cid)
+
+    const blobsToDelete = cidsToDelete.filter((cid) => !stillUsed.includes(cid))
+    await Promise.all([
+      ...blobsToDelete.map((cid) => this.blobstore.delete(CID.parse(cid))),
+      ...blobsToDelete.map((cid) => {
+        const paths = ImageUriBuilder.commonSignedUris.map((id) => {
+          const uri = this.imgUriBuilder.getCommonSignedUri(id, cid)
+          return uri.replace(this.imgUriBuilder.endpoint, '')
+        })
+        return this.imgInvalidator.invalidate(cid, paths)
+      }),
+    ])
```

Comment:
Do all these happen during the course of a transaction?  I ask because they may take a long time.  If so, maybe we could pull them out using `onCommit()` and plop them on the background queue.

Reply:
that's a good call. i'm down to move to background queue